code/modules/holomap/holomap.dm

```diff
@@ -0,0 +1,500 @@
+		M.flash_eyes(2, 1)
+	if(user in display_targets)
+		stopdisplaymap(user)
+	else
+		displaymap(user)
+
+/obj/machinery/holomap/proc/displaymap(mob/M)
+	if(!M.client)
+		return
+	display_targets |= M
+	M.client.screen += my_map.using
+	M.client.screen += my_map.arrow
+	var/datum/action/show_location/SL = new /datum/action/show_location(src)
+	actions_users[M.key] = SL
+	SL.Grant(M)
+	update_icon(UPDATE_ICON)
+	if(do_after(M, 5 MINUTES, FALSE, src, FALSE))
+		atom_say("<span class='warning'>Maximum display time reached, shutting down session.</span>")
+		stopdisplaymap(M)
+	else
+		stopdisplaymap(M)
+
+/obj/machinery/holomap/proc/stopdisplaymap(mob/user)
+	display_targets -= user
+	user.client.screen -= my_map.using
+	user.client.screen -= my_map.arrow
+	var/datum/action/show_location/SL
+	if(actions_users[user.key])
+		SL = actions_users[user.key]
+		user.client.screen -= SL.location_displayed
+		SL.Remove(user)
+	actions_users -= user.key
+	if(SL)
+		qdel(SL)
+	update_icon(UPDATE_ICON)
+
+/obj/machinery/holomap/update_icon()
+	. = ..()
+
+	if(stat & (NOPOWER|BROKEN))
+		icon_state = "map_off"
+		return
+
+	if(length(display_targets))
+		icon_state = "map_projecting"
+	else if(emagged)
+		icon_state = "map_emag"
+	else
+		icon_state = "map"
+
+/obj/machinery/holomap/emag_act(user as mob)
+	emagged = TRUE
+	to_chat(user, "You short out the emission regulators of [src]!")
+	for(var/mob/living/carbon/M in display_targets)
+		M.flash_eyes(2, 1)
+
+/obj/machinery/holomap/power_change()
+	. = ..()
+	update_icon()
+
+/obj/screen/holomap/proc/make_map()
+	var/icon/bigassicon = icon('icons/effects/480x480.dmi',"blank")
+	/*
+		We're using a 480x480 icon since thats 15x15 32x32 tiles (default view range)
+		Since maps aren't that (they are 255x255) the compromise is defaulting to render
+		starting from [9,9] up to [248,248] which is the center portion of the map with
+		a border of 8 tiles that get culled. This can be changed with `region_selection`;
+		Any region larger than 240 will have its edges culled to fit, regardless of the
+		region size it will be centered in the `bigassicon`. Happy Viewing.
+	*/
+	var/region_width = (region_selection[3] - region_selection[1]) + 1
+	var/region_height = (region_selection[4] - region_selection[2]) + 1
+
+	var/region_x_start = (region_width > 240) ? 9 : region_selection[1]
+	var/region_y_start = (region_height > 240) ? 9 : region_selection[2]
+
+	var/region_x_finish = (region_width > 240) ? region_x_start + 239 : region_selection[3]
+	var/region_y_finish = (region_width > 240) ? region_y_start + 239 : region_selection[4]
+
+	var/render_x_start = (region_width > 240) ? -8 : 128 - (region_width / 2)
+	var/render_y_start = (region_width > 240) ? -8 : 128 - (region_height / 2)
+
+	var/icon_to_use = 'icons/effects/holomap_parts.dmi'
+
+	for(var/x_index in region_x_start to region_x_finish)
+		for(var/y_index in region_y_start to region_y_finish)
+			var/turf/T = locate(x_index, y_index, zlevel_rendered)
```

Comment:
```suggestion
			CHECK_TICK
			var/turf/T = locate(x_index, y_index, zlevel_rendered)
```
if this proc EVER runs outside of init runlevel it will murder the MC. We need to be checking the tick for every turf we go over

Reply:
AA said earlier on this PR not to have CHECK_TICK in a proc thats called during `Initialize()` which this is.

Reply:
erghh, this is really dangerous.

All it takes is this being spawned outside of the init runlevel during the round for it to cause significant time dilation. I wonder if we can find a way to run this proc asynchronously of the init for holomaps

Reply:
They are currently only mapped in and not able to be constructed. Accounting for this seems.. excessive.